### ISSUE TITLE
Performance observations with call

### ISSUE THREAD

**Issue by PVince81** [nextcloud/spreed#4972]
## Steps to reproduce
1. Have a call with ~32 people (including self)
2. Allow microphone and camera
3. Everyone muted and all cameras off
4. Only one person speaking
5. Switch to grid view
6. Close all sidebars
7. Observe CPU usage, temperature
8. Switch to another tab in the same window as the call, do measurements
9. Tap your microphone for a minute to trigger the notification
10. Switch back to the call tab, do measurements

### Expected behaviour
No significant differences

### Actual behaviour
Significant differences and also additional repeated "setTimeout" events in short time frames when the tab is focussed.

## Talk app

**Talk app version:** 11.0.0alpha.3 (main company instance)

**Custom Signaling server configured:** yes

**Custom TURN server configured:** yes

**Custom STUN server configured:** yes

## Browser

**Microphone available:** yes, muted

**Camera available:** yes, muted

**Operating system:** openSUSE Tumbleweed 20210113

**Browser name:** Firefox

**Browser version:** 84.0.2 (64-bit)

### Browser log

<details>
```
Insert your browser log here, this could for example include:
a) The javascript console log
b) The network log
c) ...
```

</details>

## Server configuration

Nextcloud company instance (NC 21.0.0 beta6)

## Observations during call

See "Steps to reproduce" for more detailed conditions.
My hardware: Dell XPS 13 9333.
Due to overheating I had to add a stand with additional fans which were on all the time during the tests.
Turboboost was enabled.

Note: most values have been observed on sight and averaged intuitively.

### Tab is not focused

* CPU usage: 	150-170 %
* CPU speed: ~2.3 Ghz
* Avg temp: 85 degrees
* about:performance reported 2-3.5
* perf profile: https://cloud.nextcloud.com/s/NXSgNq62rQosyAw

### Tab is focused

* CPU usage: 200-210 %
* CPU speed: ~2.3 Ghz
* Avg temp: 91 degrees
* about:performance load 4-5 (but not high)
* perf profile: https://cloud.nextcloud.com/s/eLRBW9SerarqKHM

### Tapping microphone when tab not focused

* CPU usage: 170 % CPU
* about:performance load increased to 8! (could be the notification)
* perf profile: to do and analyze separately

### After leaving the call

I made sure to close the sidebars again.
* CPU ~24 % (???)
* Avg temp: forgot to measure but I could hear the fans
* about:performance low 0.4
* perf profile: https://cloud.nextcloud.com/s/ceWLRjXQzTNsyYw

### After leaving the call and page reload
* CPU: ~ 4%
* perf profile: https://cloud.nextcloud.com/s/o7ei99feEZXgSQL

**Comment by PVince81** [comment]
Today had another call but didn't do a capture.
I was still looking at temperature and CPU values and they were roughly the same.

One difference was that now there was no blur effect any more on the grid.
This means that the "no blur" optimization didn't improve performance in my Firefox+Linux instance.

**Comment by PVince81** [comment]
Today's call I had some troubles specific to my laptop.
For some reason turboboost didn't kick in, so the CPU freq of the 4 cores was stuck at 1.6 Ghz.
This resulted in stuttering audio.

I did some capture from that moment but didn't get any new information that we didn't already have above.
As before I saw a lot of DOM events of type "message" likely from the signaling server, not sure why there are so many but I couldn't inspect their contents.

After rebooting and fixing turboboost the CPU freq was stable around 2.2 Ghz (max is 3.1+) and I could leave the call tab open.
It was not necessary to switch to another tab any more.
It is likely that the new paginated grid view and the removal of blur helped improve performance there :tada:

**Comment by PVince81** [comment]
Today, after leaving the call with the usual number of people, I went to "about:memory" to find out about memory usage.

Some unrelated memory capture here:

<img width="632" alt="image" src="https://user-images.githubusercontent.com/277525/116255288-63f45780-a772-11eb-9abb-e82cd609b59d.png">

Seems it's using 1 GB of RAM with 600 GB unclassified.

But the most worrying is that the tab's PID is stuck at around 46% CPU despite nothing happening, and even when the tab is invisible / not focused.

The performance capture didn't show any high usage. 80% of the time is spent idle and the rest is in Network, GC, etc.
It's as if this is happening on a lower level in Firefox, maybe some of the WebRTC thing didn't shut down properly...

**Comment by PVince81** [comment]
about:performance doesn't show anything unusual for that tab, just the occasional high every many seconds when fetching messages. CPU usage for that process in `top` stays at 50% when doing nothing and jumps to 80% when fetching.

**Comment by PVince81** [comment]
after a page refresh the CPU usage is down to ~12 % again

**Comment by djtilo-ol** [comment]
Hi, we're using BBB in production to teach our students. For our staff meetings we have used MS Teams over the past year out of lack for alternatives, but want to switch to open source for obvious reasons. I have also set up a HPB and am running into the issues @PVince81 has mentioned with extremely high CPU load on a variety of devices, esp. Laptops/PCs (Dual-Cores are having a hard time, colleagues are moaning about poor connections etc). As I am quite deeply involved in maintaining our BBB Cluster, I have noticed, that CPU usage peaks there as soon as cameras are switched on - and it is related to webrtc at this stage that the different streams are not combined into one, as the proprietary clients of Zoom, Teams etc. do, but build up, so that 5 cameras mean that as many streams have to be handled which caused many dropouts in previous versions of BBB. Now they have introduced pagination, so that no more than 5 streams are connected on Desktops/Laptops and not more than 2 on mobile devices by default. This has solved the high CPU-load issue there and the related loss of connection.
In Spreed I have the impression, that the video is captured and streamed continuously to the, let's say 15 participants, and is causing the above mentioned high CPU load (heat and noise), but is "covered" by the Avatar/Initials as everyone is trying to save bandwidth by switching off the camera. I believe this because you can instantly switch on and off the video and it does not lag in the sense of having to initiate the stream, as it takes 0,5-1s in BBB or be it in Teams.
I'd like to know where I can really switch off the cameras, as one can do now individually in BBB.
For small groups I find Talk/Spreed more appealing.
Great work!

**Comment by PVince81** [comment]
> In Spreed I have the impression, that the video is captured and streamed continuously to the, let's say 15 participants, and is causing the above mentioned high CPU load (heat and noise), but is "covered" by the Avatar/Initials as everyone is trying to save bandwidth by switching off the camera.

Yes, this is currently the case. Also more observations here https://github.com/nextcloud/spreed/issues/4355.

I think the reason was to avoid having the connection delay you mention exists in other tools like BBB.

**Comment by schiessle** [comment]
> I think the reason was to avoid having the reconnection delay you mention exists in other tools like BBB.

I observed something similar with other solutions. E.g. Jitsi seems to disable the video if someone doesn't speak for some time / is muted. Only if the user unmute/speak the video comes back. The same I saw on Webex, there is is even uglier. If someone starts speaking after a while you see a few seconds a spinner in the video area of the speaker before the video shows up.

You could argue that all this looks really shitty, and I would agree. But at the end working calls are more important, imho.

Btw, something else I observed, not sure if it is related you your observation @PVince81 :

Calls with a large number of people (even if video is switched off) works smoother on my small laptop screen than on my large external monitor when using the grid view. On the external monitor the grid view contain way more avatars per page and I have the feeling that also just showing the avatars eats a lot of CPU. Maybe a hard limit for the grid size, independent from the screen size could help as well to reduce the load?

**Comment by nickvergessen** [comment]
I don't think the avatars add a lot. Might be more the overall screensize on your computer than.
But yeah we have several things on our theming list for the next major versions to improve those situations.

**Comment by nickvergessen** [comment]
@PVince81 you still on that maschine? In case feedback after Tuesday would be nice!

**Comment by PVince81** [comment]
> @PVince81 you still on that maschine? In case feedback after Tuesday would be nice!

It's still the same machine. At least the perf improved a lot after I cleant out the dusty fans in the laptop.

I'll keep an eye on perf next week.

**Comment by PVince81** [comment]
with 44+ people on 23.x, all with video disabled

- CPU: 60 degrees, very acceptable
- CPU freq: around 3 GHz
- Firefox CPU usage: stable around 40% with occasional spike

**Comment by jospoortvliet** [comment]
Same situation - Firefox too, on a Intel(R) Core(TM) i7-8550U CPU @ 1.80GHz (low power CPU) total system, CPU usage between 20 and 28. VERY good, way better than before!
